sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc

```diff
@@ -0,0 +1,674 @@
+  auto exe_bndl =
+    syclex::get_kernel_bundle<iota, sycl::bundle_state::executable>(ctxt);
+
+  // Get a kernel object for the "iota" function from that bundle.
+  sycl::kernel k_iota = exe_bndl.ext_oneapi_get_kernel<iota>();
+
+  float *ptr = sycl::malloc_shared<float>(NUM, q);
+  q.submit([&](sycl::handler &cgh) {
+    // Set the values of the kernel arguments.
+    cgh.set_args(3.14f, ptr);
+
+    cgh.parallel_for({NUM}, k_iota);
+  }).wait();
+}
+```
+
+=== Free function kernels which are templates or overloaded
+
+A free function kernel may be defined as a function template.
+It is also legal to define several overloads for a free function kernel.
+The following example demonstrates how to get a kernel identifier in such
+cases.
+
+```
+namespace syclex = sycl::ext::oneapi::experimental;
+
+template<typename T>
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
+void iota(T start, T *ptr) {
+  // ...
+}
+
+SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL
+void ping(float *x) {
+  // ...
+}
+
+SYCL_EXT_ONEAPI_SINGLE_TASK_KERNEL
+void ping(int *x) {
+  // ...
+}
+
+int main() {
+  // When the free function kernel is templated, pass the address of a
+  // specific instantiation.
+  sycl::kernel_id iota_float = syclex::get_kernel_id<iota<float>>();
+  sycl::kernel_id iota_int = syclex::get_kernel_id<iota<int>>();
+
+  // When there are multiple overloads of a free function kernel, use a cast
+  // to disambiguate.
+  sycl::kernel_id ping_float = syclex::get_kernel_id<(void(*)(float))ping>();
+  sycl::kernel_id ping_int = syclex::get_kernel_id<(void(*)(int))ping>();
+}
+```
+
+
+== Issues
+
+* We're pretty sure that we want to define some syntax that allows a free
+  function kernel to be enqueued using the APIs defined in
+  link:../proposed/sycl_ext_oneapi_enqueue_functions.asciidoc[
+  sycl_ext_oneapi_enqueue_functions], but we haven't settled on the exact API
+  yet.
+  One option is like this:
++
+```
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
+void iota(float start, float *ptr) { /*...*/  }
+
+int main() {
+  sycl::queue q;
+  float *ptr = sycl::malloc_shared<float>(N, q);
+  sycl::parallel_for<iota>(q, {N}, 1.f, ptr);
+}
+```
++
+Another option is like this:
++
+```
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
+void iota(float start, float *ptr) { /*...*/  }
+
+int main() {
+  sycl::queue q;
+  float *ptr = sycl::malloc_shared<float>(N, q);
+  sycl::parallel_for(q, {N}, kfp<iota>, 1.f, ptr);
+}
+```
++
+Where `kfp` would have some nicer name.
+
+* Can the front-end define a trait like this, which returns true only if the
+  address is a free function kernel:
++
+```
+template <auto *Func>
+bool is_kernel_function_v;
+```
++
+If the front-end can provide this, we can provide a nice diagnostic when the
+user passes an invalid address to the kernel bundle functions like
+`template<auto *Func> kernel_id get_kernel_id()`.
+
+* We need to investigate whether there will be problems passing kernel
+  arguments that are "decomposed" by the front-end.
```

Comment:
I think you are saying that we should prohibit kernel arguments that require decomposition, which means the application will not be able to pass either `accessor` or `local_accessor`.  Correct?  It might be reasonable to prohibit `accessor` because applications can use USM pointers instead.  However, prohibiting `local_accessor` seems like a big limitation because this is the main way to allocate SLM in a kernel.

I suppose we could require applications to use [sycl_ext_oneapi_work_group_local](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_local.asciidoc) if they need SLM, but that seems less convenient than using `local_accessor`.

Do you think we could allow both `accessor` and `local_accessor` as kernel arguments and teach `set_arg` to know that these arguments need to be decomposed into several low-level kernel arguments?  It seems like this might not be too difficult.  I think we could prohibit more complicated cases, for example, when an `accessor` is a member variable of a struct.

Reply:
> Can you define how SLM should be represented in "raw kernel"?

This depends on the outcome of this conversation.

If we agree that it is OK to pass `local_accessor` as a kernel argument, then you simply define a "raw kernel" with an argument of type `local_accessor`.  The kernel can then use any of the standard member functions of `local_accessor` to access the SLM.

Reply:
@bader and I had a meeting, and we decided that the extension will be able to support any kernel argument type, even those that require decomposition like `accessor` and `local_accessor`.  I resolved the related issue in c0f2029a3360ea105fbb6859ae98c025a9d120dd.